Fix count-items failures

francoisferrand opened the pull request:
- Properly handle cold objects which are currently restored
- Skip over blank PHD markers
- Do not call callback multiple times on process entry error
- Bump node to 16.20.2 (CVE fixes)

Issue: S3UTILS-141
Issue: S3UTILS-142

francoisferrand committed:
Fix handling of restored objects

The code was trying to find out the "cold" location of restored objects
by looking at the `value.location` field which is neither projected nor
contains that information.

In earler Zenko releases (up to 2.6.34), the information was simply
not available. In 2.6.35 however, code was updated to work similar to
AWS (reporting the cold location via the storage class), and the name of
the location was thus added to the `x-amz-storage-class` field.

Issye: S3UTILS-141

francoisferrand committed:
Bump setup-note action

Issue: S3UTILS-141

francoisferrand committed:
Ignore blank PHD marker

In some case, PHD are re-created form scratch and do not hold any
information (when deleting a delete marker which is the latest version).
This should be very transient (they should be cleaned after 5 seconds),
but should not create any error and we can just skip the entry.

We should also skip any error from `_processEntryData()`, as they should
not really happen (only corrupted DB or coding issue), and should thus
not prevent the process from completing.

Issue: S3UTILS-142

francoisferrand committed:
Bump node baseline to 16.20.2-bullseye-slim

Fix a few CVE vulnerabilities.

Issue: S3UTILS-141

francoisferrand committed:
Refactor: simplify output of _processEntryData

Issue: S3UTILS-142

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/s3utils/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-141 contains:


* `1.13.22`

* `1.14.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.14.0`


Please check the `Fix Version/s` of S3UTILS-141, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-141 contains:


* `1.13.22`

* `1.14.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.13.24`

* `1.14.0`


Please check the `Fix Version/s` of S3UTILS-141, or the target
branch of this pull request.

nicolas2bert commented on utils/S3UtilsMongoClient.js:
```diff
@@ -172,29 +183,28 @@ class S3UtilsMongoClient extends MongoClientInterface {
+            return { error: new Error('no bucket name provided') };
+        }
+
+        if (entry.value.isPHD && !entry.value.hasOwnProperty('content-length')) {
```
PHD-flagged master suggests that the master is no more considered valid. Do we need to exclude all master flagged with PHD to ensure we don't provide incorrect information to users? 

Additionally, the presence of the PHD flag might signify that the data for the old master hasn't been deleted (yet). This could explain the rationale behind retaining it.

francoisferrand replied:
> PHD-flagged master suggests that the master is no more considered valid

It is not valid in the sense that there is indeed no matching version; but there may still be a master (e.g. pointing to another version), we just don't know which one...

I think there is no easy fix, the proper way is to count only the versions [except for non versioned, version suspended...], deduce if there is a master from the list and do not try to substract the size/count of masters from versions at the end of the process as is done now. This shall be done separately though, in [ARTESCA-9329](https://scality.atlassian.net/browse/ARTESCA-9329)

@nicolas2bert what do you think we should do _just now_, as the least incorrect value : ignore all PHDs, or count them as master when they have a length?

nicolas2bert replied:
Both solutions provided are incorrect, but the approach of disregarding all PHD flags appears to be more accurate and easier to comprehend. 
This approach is more accurate because the PHD flag is removed when no version is present, whereas the metadata associated with the PHD flag is consistently unreliable or incorrect.
Additionally, this approach is easier to understand since it focuses on a single property, namely 'isPHD' instead of two: `entry.value.isPHD && !entry.value.hasOwnProperty('content-length')`.

Again, I don't have a strong opinion since both solution return incorrect results.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/S3UTILS-141` into
`development/1.13`
* [w/1.14/bugfix/S3UTILS-141](https://github.com/scality/s3utils/compare/development/1.14...w/1.14/bugfix/S3UTILS-141) will be merged into `development/1.14` 



The following branches will **NOT** be impacted:

* `development/1.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand committed:
Ignore all PHD docs

PHD has data from an outdated document (a master which got removed), so
we should just ignore them.

Issure: S3UTILS-141

francoisferrand committed:
bump version 1.13.24

Issue: S3UTILS-141

francoisferrand commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.13`

* :heavy_check_mark: `development/1.14`



The following branches will **NOT** be impacted:

* `development/1.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.13`

* :heavy_check_mark: `development/1.14`



The following branches have **NOT** changed:

* `development/1.4`



Please check the status of the associated issue S3UTILS-141.

Goodbye francoisferrand.
